sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2102,6 +2102,55 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns a new Dataset by adding columns or replacing the existing columns that has
+   * the same names.
+   */
+  private[spark] def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = {
```

Comment:
@cloud-fan should look at this `withColumns` before in #17819. cc @cloud-fan to see if you has more comments.

Reply:
ping @cloud-fan or @gatorsmile Can you check the SQL part? Thanks.

Reply:
`withColumn` can be reimplemented by calling this func?

Reply:
It can be. But even we want to do it, I'd prefer in a follow-up instead of this.

Reply:
I still think we should do it to avoid duplicate codes.

Reply:
Ok. Then I will do it in this PR.